### PR TITLE
Remove boilerplate for implementing `EventRetrieving` for contracts

### PR DESCRIPTION
**Pull request by nlordell**
Follow up to #606.

This PR addresses comments from the previous PR, specifically:
- Implements a declarative macro to implement `EventRetrieving` for contracts bindings generated by `ethcontract`.
- Removes the `web3()` function from the `EventRetrieving` trait, requiring a `BlockRetrieving` to be passed when constructing event handlers.

With help from @bh2smith!

### Test Plan

No logic changes, just refactoring (and moveing small amounts of code around). CI should continue to pass.

### CODE REVIEW

**Review by vkgnosis — APPROVED** [review]

File: contracts/build.rs [hunk 1]
```diff
@@ -55,13 +55,14 @@ fn main() {
...
     generate_contract("SushiswapV2Pair", hashmap! {});
-    generate_contract(
+    generate_contract_with_config(
         "GPv2Settlement",
         hashmap! {
             1 => (Address::from_str("0x3328f5f2cEcAF00a2443082B657CedEAf70bfAEf").unwrap(), Some("0x34b7f9a340e663df934fcc662b3ec5fcd7cd0c93d3c46f8ce612e94fff803909".parse().unwrap())),
             4 => (Address::from_str("0x3328f5f2cEcAF00a2443082B657CedEAf70bfAEf").unwrap(), Some("0x52badda922fd91052e6682d125daa59dea3ce5c57add5a9d362bec2d6ccfd2b1".parse().unwrap())),
             100 => (Address::from_str("0x3328f5f2cEcAF00a2443082B657CedEAf70bfAEf").unwrap(), Some("0x95bbefbca7162435eeb71bac6960aae4d7112abce87a51ad3952d7b7af0279e3".parse().unwrap())),
         },
+        |builder| builder.with_contract_mod_override(Some("gpv2_settlement")),
```
Comment (bh2smith) [inline contracts/build.rs:65]: Awesome! Now we don't have to look at `g_pv_2_settlement` anymore!

**Review by bh2smith — APPROVED** [review]
Wicked sweet! Macro Man

🇲🇰

File: contracts/build.rs (hunk 1 above)
Comment (nlordell) [reply contracts/build.rs:65]: It bothered me too much :stuck_out_tongue:

File: shared/src/event_handling.rs [hunk 2]
```diff
@@ -190,11 +190,11 @@ impl BlockNumber {
...
 #[macro_export]
 macro_rules! impl_event_retrieving {
-    ($vis:vis $name:ident for $contract_module:ident) => {
-        $vis struct $name(self::$contract_module::Contract);
+    ($vis:vis $name:ident for $($contract_module:tt)*) => {
+        $vis struct $name($($contract_module)*::Contract);
 
         impl ::shared::event_handling::EventRetrieving for $name {
-            type Event = self::$contract_module::Event;
+            type Event = $($contract_module)*::Event;
```
Comment (bh2smith) [inline shared/src/event_handling.rs]: Wowie you found it! Where/how?

Comment (nlordell) [reply shared/src/event_handling.rs]: I gave up and used `tt` (which is kind of a "match everything").

File: contracts/build.rs [hunk 3]
```diff
@@ -76,6 +77,14 @@ fn main() {
...
     name: &str,
     deployment_overrides: HashMap<u32, (Address, Option<TransactionHash>)>,
+) {
+    generate_contract_with_config(name, deployment_overrides, |builder| builder)
+}
+
+fn generate_contract_with_config(
+    name: &str,
+    deployment_overrides: HashMap<u32, (Address, Option<TransactionHash>)>,
+    config: impl FnOnce(Builder) -> Builder,
```
Comment (fleupold) [inline contracts/build.rs:87]: Nbd, but maybe deployment overrides and remaining config should be handled the way?

**Review by fleupold — APPROVED** [review]
Three nice little imrpovements...

File: contracts/build.rs (hunk 3 above)
Comment (nlordell) [reply contracts/build.rs]: Great idea! Turned out quite nice I think :smile:

**Pull request merged**